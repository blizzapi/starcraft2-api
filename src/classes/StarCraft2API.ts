import {
  BlizzAPI,
  RegionIdAsNumberOrString,
  Locale,
  QueryOptions,
} from 'blizzapi';
import * as helpers from '../helpers';
import {
  PlayerObject,
  League,
} from '../types';

export class StarCraft2API extends BlizzAPI {
  queryStaticProfileData(
    regionId: RegionIdAsNumberOrString,
    locale?: Locale,
    options?: QueryOptions,
  ): Promise<object> {
    const queryLocale = locale || BlizzAPI.getDefaultLocaleNameForRegionId(regionId);
    return this.query(
      `/sc2/static/profile/${regionId}?locale=${queryLocale}`,
      options,
    );
  }

  queryProfileMetadata(
    playerObject: PlayerObject,
    locale?: Locale,
    options?: QueryOptions,
  ): Promise<object> {
    const { regionId, realmId, profileId } = playerObject;
    const queryLocale = locale || BlizzAPI.getDefaultLocaleNameForRegionId(regionId);
    return this.query(
      `/sc2/metadata/profile/${regionId}/${realmId}/${profileId}?locale=${queryLocale}`,
      options,
    );
  }

  queryProfile(
    playerObject: PlayerObject,
    locale?: Locale,
    options?: QueryOptions,
  ): Promise<object> {
    const { regionId, realmId, profileId } = playerObject;
    const queryLocale = locale || BlizzAPI.getDefaultLocaleNameForRegionId(regionId);
    return this.query(
      `/sc2/profile/${regionId}/${realmId}/${profileId}?locale=${queryLocale}`,
      options,
    );
  }

  queryLadderSummary(
    playerObject: PlayerObject,
    locale?: Locale,
    options?: QueryOptions,
  ): Promise<object> {
    const { regionId, realmId, profileId } = playerObject;
    const queryLocale = locale || BlizzAPI.getDefaultLocaleNameForRegionId(regionId);
    return this.query(
      `/sc2/profile/${regionId}/${realmId}/${profileId}/ladder/summary?locale=${queryLocale}`,
      options,
    );
  }

  queryPlayerLadder(
    playerObject: PlayerObject,
    ladderId: number | string,
    locale?: Locale,
    options?: QueryOptions,
  ): Promise<object> {
    const { regionId, realmId, profileId } = playerObject;
    const queryLocale = locale || BlizzAPI.getDefaultLocaleNameForRegionId(regionId);
    return this.query(
      `/sc2/profile/${regionId}/${realmId}/${profileId}/ladder/${ladderId}?locale=${queryLocale}`,
      options,
    );
  }

  queryGrandmasterLeaderboard(
    regionId: RegionIdAsNumberOrString,
    options?: QueryOptions,
  ): Promise<object> {
    return this.query(
      `/sc2/ladder/grandmaster/${regionId}`,
      options,
    );
  }

  queryLeagueData(league: League, options?: QueryOptions): Promise<object> {
    const {
      seasonId,
      queueId,
      teamType,
      leagueId,
    } = league;
    return this.query(
      `/data/sc2/league/${seasonId}/${queueId}/${teamType}/${leagueId}`,
      options,
    );
  }

  querySeason(regionId: RegionIdAsNumberOrString, options?: QueryOptions): Promise<object> {
    return this.query(
      `/sc2/ladder/season/${regionId}`,
      options,
    );
  }

  queryPlayerAccount(accountId: number | string, options?: QueryOptions): Promise<object> {
    return this.query(
      `/sc2/player/${accountId}`,
      options,
    );
  }

  queryLegacyProfile(playerObject: PlayerObject, options?: QueryOptions): Promise<object> {
    const {
      regionId,
      realmId,
      profileId,
    } = playerObject;
    return this.query(
      `/sc2/legacy/profile/${regionId}/${realmId}/${profileId}`,
      options,
    );
  }

  queryLegacyLadders(playerObject: PlayerObject, options?: QueryOptions): Promise<object> {
    const {
      regionId,
      realmId,
      profileId,
    } = playerObject;
    return this.query(
      `/sc2/legacy/profile/${regionId}/${realmId}/${profileId}/ladders`,
      options,
    );
  }

  queryLegacyMatchHistory(playerObject: PlayerObject, options?: QueryOptions): Promise<object> {
    const {
      regionId,
      realmId,
      profileId,
    } = playerObject;
    return this.query(
      `/sc2/legacy/profile/${regionId}/${realmId}/${profileId}/matches`,
      options,
    );
  }

  queryLegacyLadder(
    regionId: RegionIdAsNumberOrString,
    ladderId: number | string,
    options?: QueryOptions,
  ): Promise<object> {
    return this.query(
      `/sc2/legacy/ladder/${regionId}/${ladderId}`,
      options,
    );
  }

  queryLegacyAchievements(
    regionId: RegionIdAsNumberOrString,
    options?: QueryOptions,
  ): Promise<object> {
    return this.query(
      `/sc2/legacy/data/achievements/${regionId}`,
      options,
    );
  }

  queryLegacyRewards(
    regionId: RegionIdAsNumberOrString,
    options?: QueryOptions,
  ): Promise<object> {
    return this.query(
      `/sc2/legacy/data/rewards/${regionId}`,
      options,
    );
  }

  static getAllProfileUrlLocales = helpers.getAllProfileUrlLocales;

  static checkIfProfileUrlLocaleLooksValid = helpers.checkIfProfileUrlLocaleLooksValid;

  static validateProfileUrlLocale = helpers.validateProfileUrlLocale;

  static checkIfProfileUrlLooksValid = helpers.checkIfProfileUrlLooksValid;

  static validateProfileUrl = helpers.validateProfileUrl;

  static checkIfProfileIdLooksValid = helpers.checkIfProfileIdLooksValid;

  static unpackProfileUrl = helpers.unpackProfileUrl;

  static constructProfileUrl = helpers.constructProfileUrl;

  static profileUrlRegex = () => helpers.profileUrlRegex;

  static profileUrlLocaleRegex = () => helpers.profileUrlLocaleRegex;
}
