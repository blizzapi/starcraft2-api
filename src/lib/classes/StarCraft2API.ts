import blizzapi, {
  BlizzAPIOptions,
  RegionIdAsNumberOrString,
  Locale,
  Sc2Realm
} from 'blizzapi';

import * as helpers from '../helpers';

export interface StarCraft2APIOptions extends BlizzAPIOptions {}

export interface League {
  seasonId: number | string,
  queueId: number | string,
  teamType: number | string,
  leagueId: number | string,
}

export interface PlayerObject {
  regionId: RegionIdAsNumberOrString;
  realmId: Sc2Realm;
  profileId: number | string;
  locale?: string;
}

export default class StarCraft2API extends blizzapi {
  constructor(options: StarCraft2APIOptions) {
    super(options);
  }

  queryStaticProfileData(regionId: RegionIdAsNumberOrString, locale?: Locale): Promise<object> {
    const queryLocale = locale || blizzapi.getDefaultLocaleNameForRegionId(regionId);
    return this.query(`/sc2/static/profile/${regionId}?locale=${queryLocale}`);
  }

  queryProfileMetadata(playerObject: PlayerObject, locale?: Locale): Promise<object> {
    const { regionId, realmId, profileId } = playerObject;
    const queryLocale = locale || blizzapi.getDefaultLocaleNameForRegionId(regionId);
    return this.query(
      `/sc2/metadata/profile/${regionId}/${realmId}/${profileId}?locale=${queryLocale}`,
    );
  }

  queryProfile(playerObject: PlayerObject, locale?: Locale): Promise<object> {
    const { regionId, realmId, profileId } = playerObject;
    const queryLocale = locale || blizzapi.getDefaultLocaleNameForRegionId(regionId);
    return this.query(`/sc2/profile/${regionId}/${realmId}/${profileId}?locale=${queryLocale}`);
  }

  queryLadderSummary(playerObject: PlayerObject, locale?: Locale): Promise<object> {
    const { regionId, realmId, profileId } = playerObject;
    const queryLocale = locale || blizzapi.getDefaultLocaleNameForRegionId(regionId);
    return this.query(
      `/sc2/profile/${regionId}/${realmId}/${profileId}/ladder/summary?locale=${queryLocale}`,
    );
  }

  queryPlayerLadder(playerObject: PlayerObject, ladderId: number | string, locale?: Locale): Promise<object> {
    const { regionId, realmId, profileId } = playerObject;
    const queryLocale = locale || blizzapi.getDefaultLocaleNameForRegionId(regionId);
    return this.query(
      `/sc2/profile/${regionId}/${realmId}/${profileId}/ladder/${ladderId}?locale=${queryLocale}`,
    );
  }

  queryGrandmasterLeaderboard(regionId: RegionIdAsNumberOrString): Promise<object> {
    return this.query(`/sc2/ladder/grandmaster/${regionId}`);
  }

  queryLeagueData(league: League): Promise<object> {
    const { seasonId, queueId, teamType, leagueId } = league;
    return this.query(`/data/sc2/league/${seasonId}/${queueId}/${teamType}/${leagueId}`);
  }

  querySeason(regionId: RegionIdAsNumberOrString): Promise<object> {
    return this.query(`/sc2/ladder/season/${regionId}`);
  }

  queryPlayerAccount(accountId: number | string): Promise<object> {
    return this.query(`/sc2/player/${accountId}`);
  }

  queryLegacyProfile(playerObject: PlayerObject): Promise<object> {
    const { regionId, realmId, profileId } = playerObject;
    return this.query(`/sc2/legacy/profile/${regionId}/${realmId}/${profileId}`);
  }

  queryLegacyLadders(playerObject: PlayerObject): Promise<object> {
    const { regionId, realmId, profileId } = playerObject;
    return this.query(`/sc2/legacy/profile/${regionId}/${realmId}/${profileId}/ladders`);
  }

  queryLegacyMatchHistory(playerObject: PlayerObject): Promise<object> {
    const { regionId, realmId, profileId } = playerObject;
    return this.query(`/sc2/legacy/profile/${regionId}/${realmId}/${profileId}/matches`);
  }

  queryLegacyLadder(
    regionId: RegionIdAsNumberOrString,
    ladderId: number | string,
  ): Promise<object> {
    return this.query(`/sc2/legacy/ladder/${regionId}/${ladderId}`);
  }

  queryLegacyAchievements(regionId: RegionIdAsNumberOrString): Promise<object> {
    return this.query(`/sc2/legacy/data/achievements/${regionId}`);
  }

  queryLegacyRewards(regionId: RegionIdAsNumberOrString): Promise<object> {
    return this.query(`/sc2/legacy/data/rewards/${regionId}`);
  }

  static getAllProfileUrlLocales = helpers.getAllProfileUrlLocales;
  static checkIfProfileUrlLocaleLooksValid = helpers.checkIfProfileUrlLocaleLooksValid;
  static validateProfileUrlLocale = helpers.validateProfileUrlLocale;
  static validateProfileUrl = helpers.validateProfileUrl;
  static validateProfileId = helpers.validateProfileId;
  static unpackProfileUrl = helpers.unpackProfileUrl;
  static constructProfileUrl = helpers.constructProfileUrl;
  static profileUrlRegex = () => helpers.profileUrlRegex;
}