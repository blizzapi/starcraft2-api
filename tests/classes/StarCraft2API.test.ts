import StarCraft2API from '../../src/classes/StarCraft2API';

describe('StarCraft2API', () => {
  const player = {
    regionId: 1,
    realmId: 1,
    profileId: '123',
  };

  const regionId = 1;
  const ladderId = '123';
  const accountId = '123';

  const league = {
    seasonId: 22,
    queueId: 201,
    teamType: 0,
    leagueId: 1,
  };

  const asyncMethods = [
    {
      name: 'queryStaticProfileData',
      params: [
        1,
      ],
    },
    {
      name: 'queryProfileMetadata',
      params: [
        player,
      ],
    },
    {
      name: 'queryProfile',
      params: [
        player,
      ],
    },
    {
      name: 'queryLadderSummary',
      params: [
        player,
      ],
    },
    {
      name: 'queryPlayerLadder',
      params: [
        player,
        ladderId,
      ],
    },
    {
      name: 'queryGrandmasterLeaderboard',
      params: [
        regionId,
      ],
    },
    {
      name: 'queryLeagueData',
      params: [
        league,
      ],
    },
    {
      name: 'querySeason',
      params: [
        regionId,
      ],
    },
    {
      name: 'queryPlayerAccount',
      params: [
        accountId,
      ],
    },
    {
      name: 'queryLegacyProfile',
      params: [
        player,
      ],
    },
    {
      name: 'queryLegacyLadders',
      params: [
        player,
      ],
    },
    {
      name: 'queryLegacyMatchHistory',
      params: [
        player,
      ],
    },
    {
      name: 'queryLegacyLadder',
      params: [
        regionId,
        ladderId,
      ],
    },
    {
      name: 'queryLegacyAchievements',
      params: [
        regionId,
      ],
    },
    {
      name: 'queryLegacyRewards',
      params: [
        regionId,
      ],
    },
  ];

  const staticMethods = [
    'checkIfProfileIdLooksValid',
    'getAllProfileUrlLocales',
    'checkIfProfileUrlLocaleLooksValid',
    'validateProfileUrlLocale',
    'checkIfProfileUrlLooksValid',
    'validateProfileUrl',
    'unpackProfileUrl',
    'constructProfileUrl',
    'profileUrlRegex',
    'profileUrlLocaleRegex',
  ];

  const sc2api = new StarCraft2API({
    region: 'us',
    clientId: 'sample_client_id',
    clientSecret: 'sample_client_secret',
  });

  jest.spyOn(sc2api, 'query').mockImplementation(endpoint =>
    Promise.resolve({
      data: `sample data for ${endpoint}`,
    }));

  it('should be defined', () => {
    expect(StarCraft2API).toBeDefined();
  });

  asyncMethods.forEach(asyncMethod => {
    const {
      name,
      params,
    } = asyncMethod;
    it(`should expose ${name} method`, () => {
      const methodFn = (sc2api as { [key: string]: any })[name];
      expect(methodFn).toBeDefined();
    });

    it(`${name} should match snapshot`, async () => {
      expect.assertions(1);
      const data = await (sc2api as { [key: string]: any })[name](...params);
      expect(data).toMatchSnapshot();
    });
  });

  staticMethods.forEach(staticMethod => {
    it(`should expose ${staticMethod} static method`, () => {
      const methodFn = (StarCraft2API as { [key: string]: any })[staticMethod];
      expect(methodFn).toBeDefined();
    });
  });
});
