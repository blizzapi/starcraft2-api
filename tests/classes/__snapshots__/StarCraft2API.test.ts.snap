// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarCraft2API profileUrlLocaleRegex should match snapshot 1`] = `
Object {
  "default": /\\^\\(\\?:\\[a-zA-Z\\]\\{2\\}-\\[a-zA-Z\\]\\{2\\}\\)\\$/i,
}
`;

exports[`StarCraft2API profileUrlRegex should match snapshot 1`] = `
Object {
  "default": /starcraft2\\\\\\.com\\\\/\\(\\[A-z\\]\\{2\\}-\\[A-z\\]\\{2\\}\\)\\\\/profile\\\\/\\(1\\|2\\|3\\|5\\{1\\}\\)\\\\/\\(1\\|2\\{1\\}\\)\\\\/\\(\\[0-9\\]\\\\d\\*\\)/i,
}
`;

exports[`StarCraft2API queryGrandmasterLeaderboard should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/ladder/grandmaster/1",
}
`;

exports[`StarCraft2API queryLadderSummary should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/profile/1/1/123/ladder/summary?locale=en_US",
}
`;

exports[`StarCraft2API queryLeagueData should match snapshot 1`] = `
Object {
  "data": "sample data for /data/sc2/league/22/201/0/1",
}
`;

exports[`StarCraft2API queryLegacyAchievements should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/legacy/data/achievements/1",
}
`;

exports[`StarCraft2API queryLegacyLadder should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/legacy/ladder/1/123",
}
`;

exports[`StarCraft2API queryLegacyLadders should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/legacy/profile/1/1/123/ladders",
}
`;

exports[`StarCraft2API queryLegacyMatchHistory should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/legacy/profile/1/1/123/matches",
}
`;

exports[`StarCraft2API queryLegacyProfile should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/legacy/profile/1/1/123",
}
`;

exports[`StarCraft2API queryLegacyRewards should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/legacy/data/rewards/1",
}
`;

exports[`StarCraft2API queryPlayerAccount should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/player/123",
}
`;

exports[`StarCraft2API queryPlayerLadder should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/profile/1/1/123/ladder/123?locale=en_US",
}
`;

exports[`StarCraft2API queryProfile should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/profile/1/1/123?locale=en_US",
}
`;

exports[`StarCraft2API queryProfileMetadata should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/metadata/profile/1/1/123?locale=en_US",
}
`;

exports[`StarCraft2API querySeason should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/ladder/season/1",
}
`;

exports[`StarCraft2API queryStaticProfileData should match snapshot 1`] = `
Object {
  "data": "sample data for /sc2/static/profile/1?locale=en_US",
}
`;
